node/logging-and-metrics-for-user-code.md

```diff
@@ -0,0 +1,184 @@
+For Activities, there is not an issue because users can run anything in an Activity. In Workflows OTOH, only deterministic code can run and metrics and logs should usually be dropped when a Workflow is replaying. Additionally the fact that the Node SDK runs Workflows in v8 isolates, mean that there's no way to do IO or anything that interacts with the world without explicit support from the SDK.
+
+### IO in Workflow code
+
+There are a few options for doing IO inside a Workflow's isolate:
+
+1. Load a binary (c++) module into the isolate, the module is not isolated the way JS code is
+2. Inject a function reference from the main NodeJS isolate into a Workflow isolate
+3. Accumulate data in the isolate (e.g when collecting commands to complete an activation) and pass it into a function in the main NodeJS isolate
+
+(1) is automatically ruled out because users' existing tools are probably written in JS and depend on the NodeJS environment.<br/>
+(2) has some overhead (cross thread communication and possibly blocking the isolate thread) - from the [isolated-vm](https://github.com/laverdet/isolated-vm#api-documentation) documentation:
+
+> Calling the synchronous functions will block your thread while the method runs and eventually returns a value. The asynchronous functions will return a Promise while the work runs in a separate thread pool.<br/>
+> ... <br/>
+> Additionally, some methods will provide an "ignored" version which runs asynchronously but returns no promise. This can be a good option when the calling isolate would ignore the promise anyway, since the ignored versions can skip an extra thread synchronization. Just be careful because this swallows any thrown exceptions which might make problems hard to track down.
+
+If we go for (2) we will probably recommend using the "ignored" version to reduce the extra overhead and avoid non-deterministic code from interfering with Workflow code.<br/>
+(3) has the lowest overhead but also means there's an inherent delay from logs and metrics generation to processing.
+
+### Proposed solution
+
+Allow users to expose their logging and metrics interfaces to Workflow code as external dependencies.
+We'll use either technique (2) or (3) stated above after receiving more input and running a benchmark.
+
+#### Definitions
+
+Logging and metrics require Workflow execution context to have any meaning.
+We define the `WorkflowInfo` interface which is accessible in Workflow code via `Context.info` and passed into users' "external dependency" functions.
+
+`@temporalio/workflow`
+
+```ts
+export interface WorkflowInfo {
+  workflowId: string;
+  runId: string;
+  isReplaying: boolean;
+  // TBD - more attributes
+}
+```
+
+### Logger injection example
+
+#### Variant A - `Context.dependency` + `Worker.inject`
+
+`interfaces/logger.ts`
+
+```ts
+export interface Logger {
+  info(message: string): void;
+  error(message: string): void;
```

Comment:
These functions signatures should probably more closely match the `console.log` versions (even if we aren't explicitly aping console.log) where you pass a variable amount of objects or string with substitutions